# How to generate homology models in Rosetta
## Ed Van Bruggen, Kateka Seth

## Introduction
The Homology modeling process has {{}} steps:
1. Format protein sequence
2. Generate templates
## 1. Formatting protein sequences
The first step in homology modeling is to make a FASTA file for the target protein sequence. This is simple and can be done manually. An example is shown below. Change protein_name to the name of the target protein. This should match the name of the file (eg 5vnv.fasta).

```>protein_name
EVQLQASGGGFVQPGGSLRLSCAASGSTSRQYDMGWFRQAPGKEREFVSAISSNQDQPPY
ADSVKGRFTISRDNSKNTVYLQMNSLRAEDTATYYCAFKQHHANGAYWGQGTQVTVSS```

## 2. Generating templates
The next step is to get 2-5 PDBs of similar proteins that will act as the template. This can be achieved using a NCBI website called [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi). Select Protein BLAST, then enter the target protein’s sequence. In the section "Choose Search Set," select the Protein Data Bank as the database. This will query their database of known protein structures and find the ones that are most similar to the targeted protein sequence. 

BLAST will return a list of protein sequences. Depending on the protein you are trying to simulate there might be many eligible structures to choose from. Rosetta will consider all the templates that you provide to be equally likely, even if there some templates are significantly less similar to the actual target sequence. As such, select proteins which have the highest percent identical, the fewest gaps, and the lowest E-value. Rosetta is less able to add in new amino acids compared to deleting or swapping them out, so it is often best to select sequences which have a lower similarity score but fewer gaps. If you see a large drop-off in either percent identical or E-value between the top 2 results and the following results, it is better to select only the top 2 results rather than including the next 3 results. 

Once the template proteins have been selected, their PDBs can be downloaded from the [Protein Data Bank website](http://www.rcsb.org/). However, this downloaded PDB will most likely contain more chains than the ones that were pinged by BLAST. Rosetta comes with a Python script called clean_pdb.py to get rid of any unwanted chains. This program will also produce a FASTA file for the template proteins which is needed in the next step.

Another website, [Clustal Omega](https://www.ebi.ac.uk/Tools/msa/clustalo/), is then used to combine and align all the template and target protein sequence FASTA files so they can be easily compared by Rosetta. Since the file output of Clustal Omega is not a reconized file type in Rosetta, we wrote a python script to convert this clustal file into a Grishin alignment file which Rosetta understands:

```
#!/usr/bin/env python
"""
Convert clustal alignment files to grishin for use in rosetta
Author: Ed van Bruggen <edvb@uw.edu>
"""

import sys
import argparse
from argparse import RawTextHelpFormatter

parser = argparse.ArgumentParser(description=__doc__, formatter_class=RawTextHelpFormatter)
parser.add_argument('--file', type=str, required=True,
                    help='input clustal alignment file')
parser.add_argument('--target', metavar='POS', type=int, default=1,
                    help='position of target protein (default: 1)')
args = parser.parse_args()

aln = open(args.file)
proteins = []

for i, line in enumerate(aln):
    if i == 0 or line == '\n' or line[0] == ' ':
        continue
    words = line.split()
    skip = 0
    for protein in proteins:
        if protein[0] == words[0]:
            protein[1] += words[1]
            skip = 1
            continue
    if not skip:
        proteins.append([words[0], words[1]])

target = proteins[args.target - 1]

for protein in proteins:
    if protein == target:
        continue
    grishin = open(target[0] + "_" + protein[0] + ".grishin", "w")
    grishin.write("## %s %s_thread\n#\nscores from program: 0\n0 %s\n0 %s\n" %
                  (target[0], protein[0], target[1], protein[1]))
				  ```

## Generating fragment files
To simplify the work Rosetta has to do for large proteins, fragment files can be utilized. Fragment files contain various motifs of protein folding for many amino acid sections in the given sequence. This allows Rosetta to test out sections of amino acids instead of examining one residue at a time. Fragment files can be generated by the Robetta server.

The Grishin alignment files and the template PDBs are used as input to Rosetta’s partial_thread executable to generate threaded PDBs which contain a mix of a single template and the target nanobody. An example command is shown below. Each template PDB should have its own threaded PDB.

```rosetta/main/source/bin/partial_thread.default.linuxgccrelease 
-in:file:fasta molxa3.fasta -in:file:alignment molxa3_6dbeA.grishin 
-in:file:template_pdb 6dbeA.pdb```


## Running the simulation
The threaded PDBs, target sequence FASTA file, target protein fragment files, and a RosettaCM XML file are then all given to the ```rosetta_scripts``` executable to generate the homology models of the target nanobody. The XML file is a script which informs Rosetta on the specifics of how to run the desired protocol. 

Here is an example command for rosetta_scripts.

```rosetta/main/source/bin/rosetta_scripts.default.linuxgccrelease 
-in:file:fasta molxa3.fasta -parser:protocol hybridize.xml 
-default_max_cycles 200 -dualspace```

Here is an example of the XML file for RosettaCM.

```<ROSETTASCRIPTS>
    <TASKOPERATIONS>
    </TASKOPERATIONS>
    <SCOREFXNS>
        <ScoreFunction name="stage1" weights="score3" symmetric="0">
            <Reweight scoretype="atom_pair_constraint" weight="1"/>
        </ScoreFunction>
        <ScoreFunction name="stage2" weights="score4_smooth_cart" symmetric="0">
            <Reweight scoretype="atom_pair_constraint" weight="0.5"/>
        </ScoreFunction>
        <ScoreFunction name="fullatom" weights="ref2015_cart" symmetric="0">
            <Reweight scoretype="atom_pair_constraint" weight="0.5"/>
        </ScoreFunction>
    </SCOREFXNS>
    <FILTERS>
    </FILTERS>
    <MOVERS>
        <Hybridize name="hybridize" stage1_scorefxn="stage1" stage2_scorefxn="stage2" fa_scorefxn="fullatom" batch="1" stage1_increase_cycles="1.0" stage2_increase_cycles="1.0" linmin_only="1">
            <Fragments three_mers="a2_3.frags" nine_mers="a2_9.frags"/>
            <Template pdb="1vhpA_thread.pdb" cst_file="AUTO" weight="1.000" />
            <Template pdb="5vnvA_thread.pdb" cst_file="AUTO" weight="1.000" />
        </Hybridize>
    </MOVERS>
    <APPLY_TO_POSE>
    </APPLY_TO_POSE>
    <PROTOCOLS>
        <Add mover="hybridize"/>
    </PROTOCOLS>
</ROSETTASCRIPTS>```

Running this command will generate one pdb. You can use ```--nstruct 300``` to run the command multiple times. 